harness/tests/experiment/pytorch/test_pytorch_trial.py

```diff
@@ -67,110 +68,114 @@ def test_require_global_batch_size(self) -> None:
     def test_onevar_single(self) -> None:
         """Assert that the training loss and validation error decrease monotonically."""
 
-        def make_workloads() -> workload.Stream:
-            trainer = utils.TrainAndValidate()
+        class TrainingMetricsCallback(pytorch.PyTorchCallback):
+            def __init__(self):
+                self.steps = 0
+                self.metrics = []
 
-            yield from trainer.send(steps=100, validation_freq=10)
-            training_metrics, validation_metrics = trainer.result()
-
-            # Check the gradient update at every step.
-            for idx, batch_metrics in enumerate(training_metrics):
+            def on_training_workload_end(
+                self, avg_metrics: typing.Dict[str, typing.Any], batch_metrics: typing.Dict[str, typing.Any]
+            ) -> None:
+                self.metrics.append(batch_metrics[0])
                 pytorch_onevar_model.OneVarTrial.check_batch_metrics(
-                    batch_metrics,
-                    idx,
+                    batch_metrics[0],
+                    self.steps,
                     metric_keyname_pairs=(("loss", "loss_exp"), ("w_after", "w_exp")),
                 )
+                self.steps += 1
 
-            for older, newer in zip(training_metrics, training_metrics[1:]):
-                assert newer["loss"] <= older["loss"]
+            def on_trial_shutdown(self) -> None:
+                for older, newer in zip(self.metrics, self.metrics[1:]):
+                    assert newer["loss"] <= older["loss"]
+
+        def build_callbacks(self) -> typing.Dict[str, pytorch.PyTorchCallback]:
+            return {"metricsCallback": TrainingMetricsCallback()}
+
+        setattr(pytorch_onevar_model.OneVarTrial, "build_callbacks", build_callbacks)
 
-        controller = utils.make_trial_controller_from_trial_implementation(
+        utils.run_pytorch_trial_controller_from_trial_impl(
             trial_class=pytorch_onevar_model.OneVarTrial,
             hparams=self.hparams,
-            workloads=make_workloads(),
             trial_seed=self.trial_seed,
+            max_batches=100,
+            min_validation_batches=10,
+            min_checkpoint_batches=sys.maxsize,
         )
-        controller.run()
 
     def test_multi_validation(self) -> None:
-        def make_workloads() -> workload.Stream:
-            trainer = utils.TrainAndValidate()
+        class ValidationMetricsCallback(pytorch.PyTorchCallback):
```

Comment:
The number of these suggests that you should maybe add it to the onevartrial directly?  at least for now.

Reply:
side note, `test_multi_validation` is probably dumb and probably worth just deleting, imo.

It's been a long time since supporting multiple validation metrics was a feature that had a chance of breaking.